server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java

```diff
@@ -0,0 +1,212 @@
+        }
+
+        return new JoinPlan(
+            newLhs,
+            newRhs,
+            plan.joinType(),
+            plan.joinCondition(),
+            plan.isFiltered(),
+            plan.isRewriteFilterOnOuterJoinToInnerJoinDone(),
+            true
+        );
+    }
+
+    private static LogicalPlan createLookup(Collect smallerRelation,
+                                            Symbol smallerRelationColumn,
+                                            Collect largerRelation,
+                                            Symbol largerRelationColumn,
+                                            NodeContext nodeCtx,
+                                            TransactionContext txnCtx) {
+
+        var lookUpQuery = new SelectSymbol(
+            new QueriedSelectRelation(
+                false,
+                List.of(smallerRelation.relation()),
+                List.of(),
+                List.of(smallerRelationColumn),
+                Literal.BOOLEAN_TRUE,
+                List.of(),
+                null,
+                null,
+                null,
+                null
+            ),
+            new ArrayType<>(smallerRelationColumn.valueType()),
+            SelectSymbol.ResultType.SINGLE_COLUMN_MULTIPLE_VALUES,
+            true
+        );
+
+        var anyEqImplementation = nodeCtx.functions().get(
+            null,
+            AnyEqOperator.NAME,
+            List.of(largerRelationColumn, lookUpQuery),
+            txnCtx.sessionSettings().searchPath()
+        );
+
+        var anyEqFunction = new Function(
+            anyEqImplementation.signature(),
+            List.of(largerRelationColumn, lookUpQuery),
+            DataTypes.BOOLEAN
+        );
+        WhereClause whereClause = largerRelation.where().add(anyEqFunction);
+        LogicalPlan largerSideWithLookup = new Collect(largerRelation.relation(), largerRelation.outputs(), whereClause);
+        LogicalPlan smallerSideIdLookup = new Collect(smallerRelation.relation(), List.of(smallerRelationColumn), smallerRelation.where());
+        smallerSideIdLookup = new RootRelationBoundary(smallerSideIdLookup);
+
+        return MultiPhase.createIfNeeded(Map.of(smallerSideIdLookup, lookUpQuery), largerSideWithLookup);
+    }
+
+    private static boolean applyOptimization(long lhsNumDocs, long rhsNumDocs) {
+        if ((lhsNumDocs == -1L || rhsNumDocs == -1L) ||
+            (lhsNumDocs == 0 || rhsNumDocs == 0) ||
+            lhsNumDocs == rhsNumDocs) {
+            return false;
+        }
+        long smaller, larger;
+        if (lhsNumDocs > rhsNumDocs) {
+            larger = lhsNumDocs;
+            smaller = rhsNumDocs;
+
+        } else {
+            larger = rhsNumDocs;
+            smaller = lhsNumDocs;
+        }
+        // Benchmarks revealed that the imbalance should be at least factor 10 with the larger table at least 10.000 rows
+        // See https://github.com/crate/crate/pull/15664 for details
+        if (larger < 10_000) {
```

Comment:
shouldn't we also check that the smaller table is not > 50.000 (for the benchmarks)?

Reply:
The idea here is that the larger table is at 10.000 and the smaller table 10 times smaller. So the 50.000 case is covered then, no ?

Reply:
If I understand correctly from the benchmarks, when the small table reaches 100.000, then the lookup joins becomes slower, no?

Reply:
This is a benchmark 50.000 vs 500.000 from master against this branch:

Setup:
```
create table t1 (id int primary key,a text,b text);
insert into t1 (id,a,b) select b,(b%300)::TEXT,(b%450)::TEXT from generate_series(1,500000) a(b);
create table t2 (ts timestamp,id int,c double);
insert into t2(ts,id,c) select now(),b%7000,b%6300 from generate_series(1,50000) a(b);
refresh table t1, t2;
analyze;
```
Result:
```
Q: select * from doc.t1 join doc.t2 on t1.id = t2.id
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      183.414 ±   39.788 |    162.481 |    169.734 |    171.973 |    389.991 |
|   V2    |       39.972 ±   27.423 |     35.147 |     36.344 |     37.167 |    307.615 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 128.43%                           - 129.46%
```

Reply:
When the smaller table size reaches 100.000, then both tables have the same size and then the lookup-join operation adds an overhead.

Reply:
Here is a benchmark i did which runs the query up to one million rows on each side.

![benchmark_1_million](https://github.com/crate/crate/assets/338234/bfbed5a0-12ac-4d79-8f10-dbb0c6d99aff)